### PR TITLE
Remove on_startup, on_shutdown, event system and deprecated lifespans

### PR DESCRIPTION
**Pull request by adriangb**
#2067 

Close to 500 LOC removed. That's ~ 1/20th of the codebase.

### CODE REVIEW

**Comment by euri10** [comment]
what do you think about adding a little "migration" guide like:

you used to do this:
```
async def startup():
	await open_some_ressource()
async def shutdown():
	await close_some_ressource()
app = Starlette(
	routes=routes,
	on_startup=[startup],
	on_shutdown=[shutdown]
)
```

now do this:
```
@asynccontextmanager
async def lifespan():
	try:
		await open_some_ressource()
		yield
	finally:
		await close_some_ressource()
app = Starlette(
    routes=routes,
	lifespan=lifespan
)
```

or it's too obvious ?

File: docs/database.md [hunk 1]
```diff
@@ -44,6 +45,14 @@ notes = sqlalchemy.Table(
 
 database = databases.Database(DATABASE_URL)
 
+@contextlib.asynccontextmanger
+async def lifespan(app):
+    await database.connect()
+    try:
```
Comment (euri10) [inline docs/database.md:51]: shouldn't those 2 lines be interverted ?

**Comment by adriangb** [comment]
I don't think it's too obvious, I agree we should add it. But we should probably do the same for all of the things we are deprecating / removing. Maybe it's time to start a `v1migration.md` doc that I can add to in this PR?

File: docs/events.md [hunk 2]
```diff
@@ -1,44 +1,9 @@
...
-
-Starlette will not start serving any incoming requests until all of the
-registered startup handlers have completed.
-
-The shutdown handlers will run once all connections have been closed, and
-any in-process background tasks have completed.
-
-A single lifespan `asynccontextmanager` handler can be used instead of
-separate startup and shutdown handlers:
+Starlette will not start serving any incoming requests until the lifespan has been run.
```
Comment (euri10) [inline docs/events.md:6]: you mean uvicorn right ?

File: docs/events.md [hunk 3]
```diff
@@ -64,38 +29,10 @@ for managing asynchronous tasks.
 
 ## Lifespan State
 
-The event handlers also accept a `state` argument, which is a dictionary
+The lifespan can also accept a `state` argument, which is a dictionary
 that can be used to share the objects between the startup and shutdown handlers,
 and the requests.
```
Comment (euri10) [inline docs/events.md:35]: seems to me that this becomes outdated since there wont be startup and shutdown handlers

**Comment by gnat** [comment]
So, what do we do now? This is a breaking change.


> ```
> @asynccontextmanager
> async def lifespan():
> 	try:
> 		await open_some_ressource()
> 		yield
> 	finally:
> 		await close_some_ressource()
> app = Starlette(
>     routes=routes,
> 	lifespan=lifespan
> )
> ```
This is really not obvious.

Is there any way to get the equivalent functionality of on_startup, on_shutdown more intuitively? 

The DX of being able to just specify a function for these important events is a good thing.

**Comment by gnat** [comment]
Side "meta" note, what is with all the breaking PR's this year? Is the plan to force everyone to completely re-learn Starlette and re-write their apps for 1.0... :shrug: What makes this more egregious is the proposed patterns have worse DX

@tomchristie

**Comment by adriangb** [comment]
Your example is very contrived, I seriously doubt anyone is doing that.

**Comment by gnat** [comment]
Not my code, re: https://github.com/encode/starlette/pull/2068#issuecomment-1457667798

**Comment by adriangb** [comment]
Misread. Thought you were yielding a something, which, indeed, would be breaking, but like I said very contrived.

We have documentation for this. And there are good reasons why this approach is much better than two separate functions. I’m afraid I do not agree that it is very non-obvious how to use.

**Comment by gnat** [comment]
>We have documentation for this.

Is this what you're referring to? 
* https://github.com/encode/starlette/pull/2068/files#diff-2cee6a30a4169080ec68af68d0a8555732955725fe34fbca732dd06e3e55a4a5
* https://fastapi.tiangolo.com/advanced/events/#lifespan

File: docs/applications.md [hunk 4]
```diff
@@ -3,6 +3,7 @@ Starlette includes an application class `Starlette` that nicely ties together al
 its other functionality.
 
 ```python
+from contextlib import asyccontextmanager
```
Comment (gnat) [inline docs/applications.md:6]: `asyccontextmanager` should be `asynccontextmanager`

File: docs/applications.md [hunk 5]
```diff
@@ -25,8 +26,10 @@ async def websocket_endpoint(websocket):
     await websocket.send_text('Hello, websocket!')
     await websocket.close()
 
-def startup():
+@asyccontextmanager
```
Comment (gnat) [inline docs/applications.md:29]: `asyccontextmanager` should be `asynccontextmanager`

**Comment by adriangb** [comment]
https://www.starlette.io/lifespan/

**Comment by Kludex** [comment]
Decisions can be reversed until 1.0... But a better place to talk about it would be the discussion that was taken. I'm on my phone right now.

**Comment by Kludex** [comment]
The idea of deprecating things now is to avoid doing so on 1.0.

Feedback is always welcome, and if we made a mistake, we can revert decisions.

**Comment by gnat** [comment]
For people reading this PR, FastAPI went official with this in 0.93.0, only a few days ago: https://github.com/encode/starlette/discussions/2067#discussioncomment-5215837

Totally fair with the discussion that happened in: https://github.com/encode/starlette/discussions/2067 That's fine. The cat is out of the bag at this point.

Gonna go make the changes to my Starlette apps now.

**Comment by Kludex** [comment]
Cool.

But in any case, we are always open to take decisions back if they make sense. We don't want to harm users, and I think we try to always be transparent here.

**Comment by dineshbvadhia** [comment]
@Kludex 
As of today, the FastApi/Starlette lifespan feature results in an "async generators are unimplemented" error when compiling with mypyc. Until mypyc supports async generators, the on_startup/on_shutdown feature should be retained as it passes with mypyc.

**Comment by Kludex** [comment]
> @Kludex As of today, the FastApi/Starlette lifespan feature results in an "async generators are unimplemented" error when compiling with mypyc. Until mypyc supports async generators, the on_startup/on_shutdown feature should be retained as it passes with mypyc.

Ref.: https://github.com/mypyc/mypyc/issues/868

**Comment by Kludex** [comment]
I've applied this on https://github.com/encode/starlette/pull/2384, and given co-author to @adriangb there. FYI

**Comment by graingert** [comment]
> @Kludex As of today, the FastApi/Starlette lifespan feature results in an "async generators are unimplemented" error when compiling with mypyc. Until mypyc supports async generators, the on_startup/on_shutdown feature should be retained as it passes with mypyc.

You don't need to use async generators to create an async context manager you can just manually implement `__aenter__` and `__aexit__` on a class

**Pull request closed without merging**